File: src/types/laudo.ts
/**
 * Mirror of `src-tauri/src/models/laudo.rs`.
 * Field names use snake_case to match the JSON over the Tauri bridge.
 */

export type LaudoStatus =
  | "rascunho"
  | "revisado"
  | "exportado"
  | "assinado"
  | "arquivado";

export interface Laudo {
  id: string;
  occurrence_id: string;
  title: string;
  template_id: string;
  relative_path: string;
  status: LaudoStatus;
  created_at: string;
  updated_at: string;
  last_export_pdf: string | null;
  last_export_docx: string | null;
  /** H/I — populated by `list_laudos`: tipo da assinatura digital
   *  encontrada no .sicrodoc.
   *  Ausente quando o laudo não foi assinado ainda. */
  signature_type?: "gov_br" | "sigdocs" | "A1" | "A3" | "mock" | null;
}

export interface NewLaudoInput {
  title: string;
  template_id?: string;
}

/**
 * Payload returned by `create_laudo` / `read_laudo` and accepted by
 * `save_laudo`. The `doc` field is the full `.sicrodoc` envelope (as JSON,
 * before front-end coercion). Use `coerceSicroDoc()` from the Document
 * Engine to convert it into the typed `SicroDoc`.
 */
export interface LaudoDocPayload {
  laudo: Laudo;
  doc: unknown;
}

File: src-tauri/src/models/laudo.rs
//! Laudo domain model.
//!
//! Spike B treats the document body as an opaque JSON blob (the TipTap document
//! lives entirely on the front-end). The Rust side is responsible only for:
//!   - persisting one row per laudo in the workspace's SQLite;
//!   - writing/reading the corresponding `.sicrodoc` file (UTF-8 JSON).
//!
//! Field names use snake_case to match the TypeScript wire format in
//! `src/types/laudo.ts`.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum LaudoStatus {
    Rascunho,
    Revisado,
    Exportado,
    Assinado,
    Arquivado,
}

impl LaudoStatus {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Rascunho => "rascunho",
            Self::Revisado => "revisado",
            Self::Exportado => "exportado",
            Self::Assinado => "assinado",
            Self::Arquivado => "arquivado",
        }
    }

    pub fn parse(value: &str) -> Option<Self> {
        match value {
            "rascunho" => Some(Self::Rascunho),
            "revisado" => Some(Self::Revisado),
            "exportado" => Some(Self::Exportado),
            "assinado" => Some(Self::Assinado),
            "arquivado" => Some(Self::Arquivado),
            _ => None,
        }
    }
}

impl Default for LaudoStatus {
    fn default() -> Self {
        Self::Rascunho
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Laudo {
    pub id: Uuid,
    pub occurrence_id: Uuid,
    pub title: String,
    pub template_id: String,
    pub relative_path: String,
    pub status: LaudoStatus,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub last_export_pdf: Option<DateTime<Utc>>,
    pub last_export_docx: Option<DateTime<Utc>>,
    /// H — Tipo de assinatura digital encontrada no `.sicrodoc` (se
    /// existir). Populado pelo `list_laudos` em best-effort: lê
    /// `finalization.signature.type` quando consegue, senão fica `None`.
    /// Valores típicos: `"gov_br"`, `"A1"`, `"A3"`, `"mock"`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub signature_type: Option<String>,
}

/// Input for `create_laudo`.
#[derive(Debug, Clone, Deserialize, Default)]
pub struct NewLaudoInput {
    pub title: String,
    #[serde(default = "default_template")]
    pub template_id: String,
}

fn default_template() -> String {
    "documento_livre".to_string()
}

/// Returned by read/save commands. `doc` is the full `.sicrodoc` payload
/// (schema_version + content + metadata) — the schema is enforced on the
/// front-end by the Document Engine.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LaudoDoc {
    pub laudo: Laudo,
    pub doc: serde_json::Value,
}
